### ISSUE TITLE
media queries

### ISSUE THREAD

**Issue by tomconnors** [roman01la/cljss#7]
It doesn't look like cljss supports media queries. Would you be open to adding this feature to the lib?

**Comment by roman01la** [comment]
Yes. Also I plan to work on keyframes animation syntax.

**Comment by Jumblemuddle** [comment]
+1

**Comment by brianium** [comment]
+1

@roman01la do you accept contributions? any guidelines for contributors? how can I go about helping you implement some of these things?

great library by the way 😄

**Comment by roman01la** [comment]
@brianium Yes, sure. Send small PRs and stick to the code style of the project, that's it 🙂

Regarding CSS Media Queries I haven't decided yet on the exact syntax. I'm onto something like the following but also open for suggestions

```clojure
(defstyles widget []
  {:font-size "16px"
   :media {{:screen :only :max-width 800} {:width 260}
           {:min-width 801}               {:width 460}})
```

**Comment by brianium** [comment]
@roman01la cool! I'll take a look at the syntax some more, but it looks like a good first approach. I'll take a crack at this, and maybe we can discuss further if I hit a snag.

**Comment by roman01la** [comment]
FYI: CSS keyframes support has landed in 1.5.0

**Comment by Atsman** [comment]
Hello,

I would prefer if it will look something like this:

```clojure
(defstyles widget []
  {:font-size "12px"}
  (at-media {:min-width "768px"}
                   {:font-size "16px"}))
```

Because that would be familiar for people who already use garden. And attributes in style map should be used only for css props. They should have only one responsibility. Does that male sense?

**Comment by roman01la** [comment]
Not sure yet. It's a subject to experimentation.

**Comment by Atsman** [comment]
When do you think this feature will be released?

**Comment by brianium** [comment]
@Atsman - I'm not sure if I'm blocking this at all. @roman01la is pretty on top of things, but I did say I would take a crack at it. Not sure if he is waiting on me? I've been trying to do a little housekeeping before I get to this - i.e writing tests. Also I haven't needed this feature quite yet in my current project that leverages CLJSS - so I haven't felt a rush to do it. Am I blocking you @roman01la ?

**Comment by roman01la** [comment]
@brianium Not at all. I also don't need this at the moment.

I think I'll go with Garden syntax for this, but don't know yet when it's going to land.

**Comment by Atsman** [comment]
How do you live without media queries? Or do you use also external styles where you define them ?

**Comment by roman01la** [comment]
I'm not sure how should styles declaration looks like with `at-media`.

`defstyles` looks as a function declaration
```clojure
(defstyles heading []
  {:font-size "24px"})
```

`defstyled` looks as a var declaration
```clojure
(defstyled heading :h1
  {:font-size "24px"})
```

this doesn't look good to me
```clojure
(defstyles heading []
  {:font-size "24px"}
  (at-media {:max-width "768px"} {:font-size "36px"}))

(defstyled heading :h1
  {:font-size "24px"}
  (at-media {:max-width "768px"} {:font-size "36px"}))
```

**Comment by roman01la** [comment]
namespaced keywords could help with both problems (keep original styles definition structure and don’t interfere with props)

```clojure
(require ‘[cljss.core :as css])

(defstyles h1 []
  {:font-size “24px”
    ::css/media {{:max-width “768px”}
                          {:font-size “32px”}}})
```

**Comment by Atsman** [comment]
Hi is there any problems with `at-media` approach? I like it because, you know, in css there are two types of rules,  rule set or declaration block which consists of selector + set property: value. And at-rules, statement with an at sign @. Last variant for me seems like we trying to make simple css property to play at-rule role. And the way you pass condition and actual style feels complex. That what i don't like about it. I don't know if there any technical limitations...?

What ever decision you going to make, i want it to be clear and organic in sense of css spec. Because it is kind of superset over css. And the way how it is implemented in garden feels natural. That why i proposed it initially.

What is your opinion about that?

**Comment by esseswann** [comment]
Can't we reserve submap-keys as media query rules?
```
(defstyles h1 []
  {:font-size "24px"
   {:max-width "768px} {:font-size "32px"}})
```

**Comment by den1k** [comment]
I'm in favor of @esseswann approach. It implies that key describes the constraints for the rules without having to pull in any namespaces or refer functions. Personally, `at-media` in garden always felt ad-hoc. When I write CSS, I simply want to write hashmaps of with style, some potentially dynamic. Dealing with good ol' data lets me do all of that.

**Comment by roman01la** [comment]
@den1k @esseswann LGTM, the only concern is: is this syntax unique to media queries? Just to make sure this won't be a problem in future.

**Comment by esseswann** [comment]
@roman01la I was just writing that probably it would conflict with ideas for pseudoclasses and such

**Comment by roman01la** [comment]
@esseswann `not` pseudo could have the following syntax, which is a bit better imo
```clojure
:not {:&:last-child {}
      :&:first-child {}}
```

**Comment by den1k** [comment]
@roman01la looks good. A tradeoff is that it complicates merging styles. Since multiple hash-maps with `:not` keys would need to be deep merged. I also like:

```clojure
{[:not :last-child] {:background :red},
 [:not :first-child :h1] {:background :tomato}}
```

**Comment by roman01la** [comment]
@den1k cljss doesn’t work with dynamically constructed maps of styles. Styles definition should be declarative. Everything is processed at compile time into static and dynamic styles. Dynamic part can be only CSS values

**Comment by den1k** [comment]
@roman01la noted! @esseswann it could with nested vectors. I wrote and extensively used some utils to extend garden with vectors for combined values, e.g.

```clojure
(rewrite-styles
  [:&:after {:position 2
             :top      [:rem 20 20 20 [:rotate 50] 10] ; overwrite default unit
             :border   [1 2 :solid :red]
             :more     [:rem 2 2]
             :pointer  nil}
   [:.another {:background :red}]])

 ;; =>
 [:&:after
  {:position "2px",
   :top      "20rem 20rem 20rem rotate(50rem) 10rem",
   :border   "1px 2px solid red",
   :more     "2rem 2rem"}
  [:.another {:background "red"}]])
```

This defaults numbers to pixels, but that's just my default. It could leave them as is. I really enjoyed writing styles that way. Source: https://gist.github.com/den1k/1f6caa8eb3e8a4603aa8d03c4508f512

So @esseswann based on your example
```clojure
; :not(nth-child(2)):hover
[:not [:nth-child :2] :hover]
```

**Comment by Atsman** [comment]
Maybe original garden creator can add something here @noprompt?

**Comment by Atsman** [comment]
Also i am thinking, is it possible to have a few ways of defining media queries? I personally don't like, when something solved in multiple ways, but here that maybe a good idea. it's just a matter of tastes. And when people then try different ways, community will show what is better...

**Comment by esseswann** [comment]
@den1k This is looking nice :)
@Atsman We can parse complex keys like JSS does, don't we?

**Comment by Atsman** [comment]
@esseswann definitely we can, don't see any reasons why not.

**Comment by sherbondy** [comment]
FYI, here's a comprehensive listing of CSS At-rules:
[MDN At-rules](https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule)

Beyond `@media`, both `@supports` and the (experimental) `@document` rules both have the same sort of nested definition capability and syntax, referred to in the CSS spec as [conditional group rules](https://www.w3.org/TR/css3-conditional/#processing).

So I don't think using maps as keys to indicate `@media` specifically is the best option!

IMO, the namespaced keyword seems the most robust and future-proof.

**Comment by esseswann** [comment]
@sherbondy I kind of support these concerns, explicit is better than implicit. But I think there is a room for a little creativity here, it's very elegant in `JSS` for similar cases.

**Comment by roman01la** [comment]
Just published experimental support for Media Queries in `1.6.0-SNAPSHOT`. The syntax is specified as of CSS Media Queries Level 4 spec [here](https://github.com/roman01la/cljss/blob/master/src/cljss/media.clj)

```clojure
(defstyles header [height]
  {:height height
   :cljss.core/media {[:only :screen :and [:max-width "460px"]]
                      {:height (/ height 2)}}})
```

Currently it doesn't fully support `:or` condition and dynamic values within queries (not sure if we need it at all tbh).

Supported media types: `#{:all :print :screen :speech}`
Modifiers: `#{:not :only}`

More examples of a query:

### Boolean query
```clojure
[[:monochrome]]
```
### Simple conditional query
```clojure
[[:max-width "460px"]]
```
### Basic range query
_Supported operators for range queries `#{'= '< '<= '> '>=}`_
```clojure
[[:max-width '> "400px"]]
```
### Complex range query
```clojure
[["1200px" '>= :max-width '> "400px"]]
```

Give it a try, looking forward to get more feedback.

Happy New Year! 🎉

**Comment by Atsman** [comment]
@roman01la  hello, super great :) i am going to play with it.

**Comment by roman01la** [comment]
FYI, I've updated `1.6.0-SNAPSHOT` with improved media queries. The syntax is changed a bit to match the spec. Now you can declare multiple MQs, which makes `<media-query-list>` token. In CSS it is a comma separated query (`@media screen and (max-width: 460px), print and (color)`)

```clojure
[[:screen :and [:max-width "460px"]]
 [:print :and [:color]]]
```

**Comment by roman01la** [comment]
If there's no any complains, I'm gonna publish a stable version with Media Queries support.

**Comment by esseswann** [comment]
I played a bit, seems all right for now

**Comment by roman01la** [comment]
Just published `1.6.0` with media queries support
